### PR TITLE
Fix style injection for partial HTML

### PR DESCRIPTION
**Pull request by drwpow**
## Changes

Partial fix for #1964. When a page tree doesn’t generate `<head>` successfully, some things get omitted from the page. While this is an edge case that takes specific conditions to meet (outlined in #1964), the fallback behavior could still be better.

## Testing

Added test

## Docs

Bugfix

### CODE REVIEW

File: packages/astro/test/test-utils.js [hunk 1]
```diff
@@ -59,7 +59,8 @@ export async function loadFixture(inlineConfig) {
     build: (opts = {}) => build(config, { mode: 'development', logging: 'error', ...opts }),
     startDevServer: async (opts = {}) => {
       const devServer = await dev(config, { logging: 'error', ...opts });
-      inlineConfig.devOptions.port = devServer.port; // update port
+      config.devOptions.port = devServer.port; // update port
```
Comment (drwpow) [inline packages/astro/test/test-utils.js:62]: `inlineConfig` is the reference passed into a test, but `config` is the reference used by `.fetch()`. Fixed a flakiness bug where `fetch()` had an old port.

File: packages/astro/src/core/ssr/index.ts [hunk 2]
```diff
@@ -274,6 +269,11 @@ export async function render(renderers: Renderer[], mod: ComponentInstance, ssrO
     html = await viteServer.transformIndexHtml(viteifyURL(filePath), html, pathname);
   }
 
+  // inject <!doctype html> if missing (TODO: is a more robust check needed for comments, etc.?)
+  if (!/<!doctype html/i.test(html)) {
+    html = '<!DOCTYPE html>\n' + html;
```
Comment (drwpow) [inline packages/astro/src/core/ssr/index.ts:274]: `<!DOCTYPE` insertion saved until the very end, because we may be injecting things to the beginning of the document for partials.

File: packages/astro/src/runtime/server/index.ts [hunk 3]
```diff
@@ -384,7 +384,13 @@ export async function renderPage(result: SSRResult, Component: AstroComponentFac
...
     styles.push(renderElement('style', { props: { 'astro-style': true }, children: 'astro-root, astro-fragment { display: contents; }' }));
   }
-  return template.replace('</head>', styles.join('\n') + scripts.join('\n') + '</head>');
+
+  // inject styles & scripts at end of <head>
+  let headPos = template.indexOf('</head>');
+  if (headPos === -1) {
+    return styles.join('\n') + scripts.join('\n') + template; // if no </head>, prepend styles & scripts
+  }
+  return template.substring(0, headPos) + styles.join('\n') + scripts.join('\n') + template.substring(headPos);
```
Comment (drwpow) [inline packages/astro/src/runtime/server/index.ts:393]: This fixes Astro styles when `</head>` isn’t present

File: packages/astro/src/core/ssr/html.ts [hunk 4]
```diff
@@ -27,8 +27,9 @@ export function injectTags(html: string, tags: vite.HtmlTagDescriptor[]): string
   const lastToFirst = Object.entries(pos).sort((a, b) => b[1] - a[1]);
   lastToFirst.forEach(([name, i]) => {
     if (i === -1) {
-      // TODO: warn on missing tag? Is this an HTML partial?
-      return;
+      // if page didn’t generate <head> or <body>, guess
+      if (name === 'head-prepend' || name === 'head') i = 0;
+      if (name === 'body-prepend' || name === 'body') i = html.length;
```
Comment (drwpow) [inline packages/astro/src/core/ssr/html.ts:32]: This fixes non-Astro component styles (JSX, etc.) when `</head>` isn’t present

**Review by jonathantneal — APPROVED** [review]
As someone inspired by the source of http://diveintohtml5.info, I love the idea of Astro working with plain HTML and no `<head>`.

File: packages/astro/src/core/ssr/html.ts (hunk 4 above)
Comment (drwpow) [reply packages/astro/src/core/ssr/html.ts:32]: The `0` here means “insert at character position `0`,“ i.e. append to the beginning of the document.

We aren’t appending anything to `<body>` currently, but a fix was applied nonetheless

**Review by tony-sull — APPROVED** [review]
Looks good to me!

Always a bit nerve racking touching the final HTML with string manipulations, but the fixes make sense and test coverage looks good too 👍
